### DISCUSSION TITLE
To `.handler` or not + always use term "middleware"?

### DISCUSSION THREAD

**Discussion by brillout** [magne4000/universal-middleware#60] [General]
**Pro**: Aesthetically I like the `.handler` convention.  
**Con**: Technically `.handler.js` is just a regular `.js` file so it could be a bit misleading (I usually expect `.something.js` to be special in some way).

On a tangent, I'm thinking maybe we could/should always use the term "middleware" and not focus on the "middleware"/"handler" distinction. But I don't know, maybe it's actually helpful for the documentation / clarifying things.

Maybe it should be `.middleware.js` instead of `.handler.js`. (Unless `.handler.js` cannot be a middleware.)

Anyways, just thinking out loud. I ain't that familiar with universal-middleware. You probably thought a lot more about all this than I did. (Feel free to ignore this discussion if you don't see potential in it.)

**Comment by magne4000** [comment]
> On a tangent, I'm thinking maybe we could/should always use the term "middleware" and not focus on the "middleware"/"handler" distinction. But I don't know, maybe it's actually helpful for the documentation / clarifying things.
> 
> Maybe it should be .middleware.js instead of .handler.js. (Unless .handler.js cannot be a middleware.)

We need the distinction for documentation purpose first, indeed, but not exclusively.
All server wrappers usually have 2 variants, [`createHandler`](https://github.com/magne4000/universal-middleware/blob/24d5fc3bafb0ef3b00c0eb8067d7acb86a67d079/packages/adapter-h3/src/common.ts#L35-L44) and [`createMiddleware`](https://github.com/magne4000/universal-middleware/blob/24d5fc3bafb0ef3b00c0eb8067d7acb86a67d079/packages/adapter-h3/src/common.ts#L89-L115). This is required for 2 things: Proper typing support for most servers, and better error handling. It will also generate less bloat on some targets, but that just a side effect.

#### Why `.handler` and `.middleware` naming?
The motivation was to be able to guess that a file must be wrapped by `universal-middleware` if it appears in an entry during build, in order to keep everything in a single build process (similar to what Vike does with `+` files).

**Reply by magne4000** [reply]
Also, I would like to add that some targets do not support _middlewares_, only _handlers_ (like Next.js). Handlers are the minimal founding block of `universal-middleware`. Because of that, _middlewares_ can be considered as an upgrade on top of _handlers_ for targets that support it.

**Reply by brillout** [reply]
Sounds good.

Since it actually does add functionality (to tell `universal-middleware`'s bundler what the file is defining) I guess it makes sense :+1:
